process/rating/blockSigningRater.go

```diff
@@ -0,0 +1,90 @@
+package rating
+
+import (
+	"github.com/ElrondNetwork/elrond-go/process"
+	"github.com/ElrondNetwork/elrond-go/process/economics"
+	"github.com/ElrondNetwork/elrond-go/sharding"
+	"sort"
+)
+
+// BlockSigningRater defines the behaviour of a struct able to do ratings for validators
+type BlockSigningRater struct {
+	sharding.RatingReader
+	startRating      uint32
+	maxRating        uint32
+	minRating        uint32
+	ratingOptions    map[string]int32
+	ratingOptionKeys []string
+}
+
+//NewBlockSigningRater creates a new Rater of Type BlockSigningRater
+func NewBlockSigningRater(ratingsData *economics.RatingsData) (*BlockSigningRater, error) {
+	if ratingsData.MinRating() > ratingsData.MaxRating() {
+		return nil, process.ErrMaxRatingIsSmallerThanMinRating
+	}
+	if ratingsData.MaxRating() < ratingsData.StartRating() || ratingsData.MinRating() > ratingsData.StartRating() {
+		return nil, process.ErrStartRatingNotBetweenMinAndMax
+	}
+
+	ratingOptionKeys := make([]string, 0)
+	for key := range ratingsData.RatingOptions() {
+		ratingOptionKeys = append(ratingOptionKeys, key)
+	}
+
+	sort.Strings(ratingOptionKeys)
+
+	return &BlockSigningRater{
+		ratingOptions:    ratingsData.RatingOptions(),
+		startRating:      ratingsData.StartRating(),
+		minRating:        ratingsData.MinRating(),
+		maxRating:        ratingsData.MaxRating(),
+		ratingOptionKeys: ratingOptionKeys,
+	}, nil
+}
+
+func (bsr *BlockSigningRater) ComputeRating(ratingKey string, val uint32) uint32 {
+	newVal := int64(val) + int64(bsr.ratingOptions[ratingKey])
+	if newVal < int64(bsr.minRating) {
+		return bsr.minRating
+	}
+	if newVal > int64(bsr.maxRating) {
+		return bsr.maxRating
+	}
+
+	return uint32(newVal)
+}
+
+//GetRatingOptionKeys gets all the ratings  keys for the options
+func (bsr *BlockSigningRater) GetRatingOptionKeys() []string {
+	return bsr.ratingOptionKeys
+}
+
+//GetRating returns the Rating for the specified public key
+func (bsr *BlockSigningRater) GetRating(pk string) uint32 {
+	if bsr.RatingReader == nil {
```

Comment:
shouldn't this always exist ?

Reply:
if it can be nil sometimes, than you have to check with check.ifNil

Reply:
instead of having to check, ensure it that you have a nil implementation for the ratingReader.

Reply:
added NilRatingReader and set it on the constructor.